base/multidimensional.jl

```diff
@@ -860,21 +834,30 @@ ensure_indexable(I::Tuple{}) = ()
 # until Julia gets smart enough to elide the call on its own:
 @inline to_indices(A, I::Tuple{Vararg{Union{Integer, CartesianIndex}}}) = to_indices(A, (), I)
 # But some index types require more context spanning multiple indices
-# CartesianIndex is unfolded outside the inner to_indices for better inference
+# `CartesianIndex` is unfolded directly to `Int`s for better inference
 _to_indices1(A, inds, I1::CartesianIndex) = map(Fix1(to_index, A), I1.I)
-_cutdim(inds, I1::CartesianIndex) = IteratorsMD.split(inds, Val(length(I1)))[2]
 # For arrays of CartesianIndex, we just skip the appropriate number of inds
-_cutdim(inds, I1::AbstractArray{CartesianIndex{N}}) where {N} = IteratorsMD.split(inds, Val(N))[2]
+_cutdim(inds, ::AbstractArray{CartesianIndex{N}}) where {N} = IteratorsMD.split(inds, Val(N))[2]
```

Comment:
It'd be nice to coalesce this around `index_ndims`, too.  Obviously that's not new to this PR, but would be great to cleanup at the same time.

Reply:
Since the inference of indirect rescusion get improved, I think we can revert these change from #45672.